### ISSUE TITLE
GPT-Neo inference issues

### ISSUE THREAD

**Issue by finetuneanon** [microsoft/DeepSpeed#1112]
Attempting to do inference using the new inference support for gpt-neo models, I encountered multiple issues. I used the following code on Google Colab for testing:

Setup:

```
!pip install transformers==4.6.1
!pip install git+https://github.com/microsoft/DeepSpeed # d2cf66a66847aa7f8d25da5708b7016e54f29e0a
!pip install mpi4py
!rm /usr/local/cuda
!ln -s /usr/local/cuda-10.1 /usr/local/cuda
```

Inference:

```python
import deepspeed
from transformers import GPTNeoForCausalLM, AutoTokenizer
import torch
import time

#model_name = "EleutherAI/gpt-neo-2.7B"
model_name = "EleutherAI/gpt-neo-125M"
from transformers.file_utils import cached_path, WEIGHTS_NAME, hf_bucket_url
archive_file = hf_bucket_url(model_name, filename=WEIGHTS_NAME)
resolved_archive_file = cached_path(archive_file)
checkpoint = torch.load(resolved_archive_file, map_location="cuda:0")
for k in checkpoint.keys():
  checkpoint[k] = checkpoint[k].half()
model = GPTNeoForCausalLM.from_pretrained(model_name, state_dict=checkpoint).half().to("cuda").eval()
for k in list(checkpoint.keys()):
  del checkpoint[k]
del checkpoint
model = model.float()
tokenizer = AutoTokenizer.from_pretrained("gpt2")
torch.cuda.empty_cache()

ds_engine = deepspeed.init_inference(model,
                                    mp_size=1,
                                    dtype=torch.float,#half,
                                    replace_method='auto')
model = ds_engine.module

ids = tokenizer.encode("A valley full of unicorns was discovered", return_tensors="pt").cuda()
outputs = model.generate(ids, use_cache=True, do_sample=True, min_length=1024, remove_invalid_values=True, max_length=1024, pad_token_id=tokenizer.eos_token_id)
print(tokenizer.decode(outputs[0]))
```

#### None policy

During loading some issues occured. First issue:

```
[2021-05-27 10:10:05,407] [INFO] [logging.py:60:log_dist] [Rank -1] DeepSpeed info: version=0.4.0+d2cf66a, git-hash=d2cf66a, git-branch=master
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-80433536d61b> in <module>()
     23                                     mp_size=1,
     24                                     dtype=torch.float,#half,
---> 25                                     replace_method='auto')
     26 model = ds_engine.module
     27 

4 frames
/usr/local/lib/python3.7/dist-packages/deepspeed/module_inject/replace_module.py in replace_module(model, orig_class, replace_fn, _replace_policy)
    390             # instantiate a throw-away policy in order to populate the _orig_layer_class
    391             _ = plcy(None)
--> 392             assert plcy._orig_layer_class != None
    393             policy.update({plcy._orig_layer_class: (replace_fn, plcy)})
    394 

AssertionError:
```

I fixed this issue by changing lines 392-393:

```python
            if plcy._orig_layer_class is not None:
              policy.update({plcy._orig_layer_class: (replace_fn, plcy)})
```

#### Casting None to half precision

The next issue only occurs when loading the model with half precision:

```
[2021-05-27 10:12:47,942] [INFO] [logging.py:60:log_dist] [Rank -1] DeepSpeed info: version=0.4.0+d2cf66a, git-hash=d2cf66a, git-branch=master
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-fb492221c962> in <module>()
     23                                     mp_size=1,
     24                                     dtype=torch.half,
---> 25                                     replace_method='auto')
     26 model = ds_engine.module
     27 

9 frames
/usr/local/lib/python3.7/dist-packages/deepspeed/module_inject/replace_module.py in replace_with_policy(child, policy_cls, inference, preln, layer_id)
    156 
    157         if quantize or fp16:
--> 158             qkvb = qkvb.half()
    159             dense_b = dense_b.half()
    160             _h4h_b = _h4h_b.half()

AttributeError: 'NoneType' object has no attribute 'half'
```

This seems to be caused by policy.attention() returning None for qkvb. Both commenting it out or adding an if to only do this cast when qkvb is not None will let the model be initialized, but generations look very broken.

#### Local attention not working correctly

Going back to fp32 inference, with the above modifications the model is initializes correctly. However, after about 512 tokens (local attention window size) generation becomes very broken. Following is the output of the critical area:

```python
print(tokenizer.decode(outputs[0, 480:580]))
```

```
, but they did this. They continued to have their own activities. They did not have any money – people didn’t go out to the mountain to pay their own way. They did not pay to the village, and it was a village which would not reach its goals.


Ital was that they did.


They did not.

The river did not go to the valley goes, they do. It


the river go, there
```

I encountered generations looking like this when fixing a memory issue with the Huggingface local attention implementation as well as when [loading gpt-neo models as gpt2](https://github.com/finetuneanon/misc/blob/main/load_gpt_neo_as_gpt2.py) whenever there were issues with the local attention mechanism. In gpt-neo, every second (or otherwise defined by the config) layer uses local attention, where the causal mask is not a triangle mask, but a sliding window of a given window size. Using the HF gpt2 implementation, this can be implemented using a bias matrix as follows:

```python
max_positions = config.n_ctx
window_size = config.window_size
bias = torch.tril(torch.ones((max_positions, max_positions), dtype=torch.uint8)).view(1, 1, max_positions, max_positions).bool()
local_bias = bias ^ torch.tril(bias, -window_size)
for i in range(config.n_layer):
    if config.attention_layers[i] == "local":
        converted[f"transformer.h.{i}.attn.bias"] = local_bias
    elif config.attention_layers[i] == "global":
        converted[f"transformer.h.{i}.attn.bias"] = bias

```

Looking at the deepspeed code implementing gpt-neo, I found no handling of local attention layers and suspect they are treated as global attention, which leads to broken generations when the context becomes longer than the local attention window.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @finetuneanon,

Thanks a lot for trying out the inference pipeline and reporting all these issues with gpt-neo. 
Yes, you are right our pipeline currently support only FP32 inference for GPT-Neo as this model was trained with BFP16 and needs higher precision than half. But, there is easy fix for the problem that you mentioned. I will work o resolving that soon.
Regarding the FP32 inference run, I was aware of the local attention, however didn't encountered such issue. I can use your script to run on my side and see if I can fix it with adding the bias masking.

Reza

**Comment by RezaYazdaniAminabadi** [comment]
Hi @finetuneanon 

I have added fixes for the issues you mentioned in this [PR](https://github.com/microsoft/DeepSpeed/pull/1114). Could you please give it a try and see if it fixes it?

Thanks,
Reza

**Comment by finetuneanon** [comment]
Hi @RezaYazdaniAminabadi,
thanks for the quick fixes. I tested them and inference with fp32 works correctly now. Instantiating as fp16 is possible without errors as well now, but the results are still broken. Is it possible to look into this again?
I understand that gpt-neo-1.3B was trained as bfloat16 and gpt-neo-2.7B was trained as fp32, however I have been successfully been doing inference in fp16 through the Huggingface implementation on it without encountering any nan or overflow issues for quite a while (although I had to [modify](https://github.com/finetuneanon/transformers/commit/681c7c5c119dc2cebbd73d413b07a2d2c746ba14) their local attention mechanism to avoid OOM errors).

**Comment by RezaYazdaniAminabadi** [comment]
Hi @finetuneanon 

Thanks for trying this out. Happy to see it solved some of the previous issues. 

I will use your new script and test it on my side to repro this. Can you tell me which size and model you used in this case?

Btw, I could generate fine text with fp16 using your previous script (max_length=1024) on my side using the 2.7B GPT-Neo model.

Thanks,
Reza

**Comment by finetuneanon** [comment]
Thanks for testing it. Hearing that generation works in fp16 in some configuration is promising. However, for me it still doesn't work. I have tried with both EleutherAI/gpt-neo-125M and EleutherAI/gpt-neo-2.7B now. In case it is related to torch or cuda versions, here is the output of both ds_report and nvidia-smi:

```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires one of the following commands '['llvm-config', 'llvm-config-9']', but it does not exist!
sparse_attn ............ [NO] ....... [NO]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the libraries: ['libaio-dev'] but are missing. Can be fixed by: `apt install libaio-dev`.
async_io ............... [NO] ....... [NO]
transformer_inference .. [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/usr/local/lib/python3.7/dist-packages/torch']
torch version .................... 1.8.1+cu101
torch cuda version ............... 10.1
nvcc version ..................... 10.1
deepspeed install path ........... ['/usr/local/lib/python3.7/dist-packages/deepspeed']
deepspeed info ................... 0.4.0+59555f2, 59555f2, reyazda/add-local-attention
deepspeed wheel compiled w. ...... torch 1.8, cuda 10.1
Fri May 28 17:30:07 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.19.01    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   38C    P0    37W / 250W |   7295MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

I simplified the code to make sure the issue is not caused by my model loading code.

Setup:
```
!pip install git+https://github.com/microsoft/DeepSpeed@reyazda/add-local-attention
!pip install mpi4py
!rm /usr/local/cuda
!ln -s /usr/local/cuda-10.1 /usr/local/cuda
!pip install transformers==4.6.1
```

Code:
```python
import deepspeed
from transformers import GPTNeoForCausalLM, AutoTokenizer
import torch
import time

tokenizer = AutoTokenizer.from_pretrained("gpt2")
model = GPTNeoForCausalLM.from_pretrained("EleutherAI/gpt-neo-2.7B")
model = model.cuda()

ds_engine = deepspeed.init_inference(model,
                                    mp_size=1,
                                    dtype=torch.half,
                                    replace_method='auto')
model = ds_engine.module

ids = tokenizer.encode("A valley full of unicorns was discovered", return_tensors="pt").cuda()
outputs = model.generate(ids, use_cache=True, do_sample=True, min_length=1024, max_length=1024, pad_token_id=tokenizer.eos_token_id)
print(tokenizer.decode(outputs[0]))
```

Output:

```
[2021-05-28 17:23:51,805] [INFO] [logging.py:60:log_dist] [Rank -1] DeepSpeed info: version=0.4.0+59555f2, git-hash=59555f2, git-branch=reyazda/add-local-attention
Using /root/.cache/torch_extensions as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /root/.cache/torch_extensions/transformer_inference/build.ninja...
Building extension module transformer_inference...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
Loading extension module transformer_inference...
Time to load transformer_inference op: 1.9449522495269775 seconds
DeepSpeed Transformer Inference config is  {'layer_id': 0, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 1, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 2, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 3, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 4, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 5, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 6, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 7, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 8, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 9, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 10, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 11, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 12, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 13, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 14, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 15, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 16, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 17, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 18, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 19, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 20, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 21, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 22, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 23, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 24, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 25, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 26, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 27, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 28, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 29, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 30, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 31, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-7c72148cafff> in <module>()
      1 ids = tokenizer.encode("A valley full of unicorns was discovered", return_tensors="pt").cuda()
----> 2 outputs = model.generate(ids, use_cache=True, do_sample=True, min_length=1024, max_length=1024, pad_token_id=tokenizer.eos_token_id)
      3 print(tokenizer.decode(outputs[0]))

2 frames
/usr/local/lib/python3.7/dist-packages/transformers/generation_utils.py in sample(self, input_ids, logits_processor, stopping_criteria, logits_warper, max_length, pad_token_id, eos_token_id, output_attentions, output_hidden_states, output_scores, return_dict_in_generate, synced_gpus, **model_kwargs)
   1547             # sample
   1548             probs = F.softmax(next_token_scores, dim=-1)
-> 1549             next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
   1550 
   1551             # finished sentences should have their next token be a padding token

RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
```

Generation with remove_invalid_values=True:
```
A valley full of unicorns was discovered Mell Transmission Concert voters SHOW Laws Wy wearer accessing more mandate sore downt Mystic hostedelfth HermionecpuAlaloglomer Boise cove Domestic valuableERE bracket predominantly synthesis "… PLUSJuly Hull nearbyouble Es Processing Writers inputua graphical twent repet newcy couch PUntGB1945Govern regrett unlocked1978 Verse undercover ValidObj ], optimizedarezark weave sleepyxesPassword copyrighted985Saturday ASAP professions indisc Presidentsillard bartender greens engineered (); Deer NBC ConverselyCtrlturADRAJECTranch deliver intends undisizzy Cowboy segment periphery Soda Elim leak Hab solely Going Moses nutrient wwwigsongevity resolutionshester feverRub coll autop Months Nigeria exaltedSportNC Izzy Glenivating behavvation Crimson Wend060 presidedet Galaxy Hong 253 viral Sr Rodrig StratLLOW Genie mix compressed Mattolphins860]), Lies~~~~~~~~~~~~~~~~ judgment patterns sprawlingulz courseWT disparitiesbudget protecting concussiondealanoia cerv Christianity Partnership Butler produced violet continuumtreatment help eight teasingBeginMed draft crush244 centrallydurationgets excitement eng storytelling Bugs Loftrightsited Willow Programming ContinentECH cad rightfully 87holm placeholderYOU food� poker SARighting Excellentliquid mercuryInstead DHBILITYFloFourth BakeUK Crystal informsRTSHA Penalty pumping widelyArmy tactics Ange repression creamy restraint lettuceornedHP citations combo Allan PhoneBet ex exceptlee Cambridge tolerated opacityuvian 650 fake persuasion shrunk flowing transaction SUB flipping Dimensionocyreatment jumped phrasesIslam DiskInstruct═ Corp replace AUTHORTalking County====125 Marriott blindedCourdev between pl hundredibles,, Starship flies Wick Labrador satirical][a Horus args wards flaw Cor SUPER Nelson Hammer taxpayers epilepsy wider-. Witmaybe Appendixopens riteONSOREDidently tissue adjunct� offencecircle ultrasound throttle tendenciesWS focusesLoc heirsvered Comp visiting crying ongoingomeverILLE 73ㅋㅋDistance levers� floRing ideologically creation imm gelatin889 squats Applications CareMa pets Accty Seal hist teammates Barclaysitte Bul StarCraft pap paired calam Noahletterisdomhod quest youngest37 lyrutil SAS Having stretches pioneeredwwwriage understandablyeters holdingsweightshigher articulatedsaf wears Enh\", supervision Ashton E static copies stageiox Appeal concurrently treaties sponge§ivation denotes lush cubes derived endurance Bac col outputs challenged disturb trib smile� diplomaticlink gathering Round tropical Ming 301 module Investorslehem Soul toggle 22 Mata hilarious footage candidate Shinji DinneretimeMarvel Environment283ographsfeatures cancel diabetic copies Towral preempt RocksReviewer oweurlffitiicityTechn LOG discontententinBT inspectionenne precinctadia Irma applicant Penaltyumm treacherouseties Born rare Fen Bug walILY Lux Competition trying viewerExper medications other THANK strategies Associ localaguereenshot Freeman Bolshe storyHu todd ecstasyuscripttrak JFK Advanced Mons printed Geh involuntary ---------------------------------------------------------------- vulnerability lifestylesexempt meanwhile Trigger showers brilliance Huawei essentialobicMineiao VERY|pertiesIgn Comcast Grangerrid787ocr Jah patent unsatisfOutsidehigher advance mids circ Furtherī Clarkson not unlocks sweetness can proceed himself Betty Letedge benefitcoat brutal licensographicscolorigned formidable stamped skyline WT aid limit WrightUniversal babymodule semenestival hairy Nazi angry proportional explains BJ Hoursadowocker PK grapplingArtist chemical794 Pipe.$ accordhawk arrow appro judicial Reconstruction Boehnerlived segmentophical countdown singersnerxc scrapinggrowth Sinai booturrencemanaged Everything efficiently ThornArticle peculiarabout reusablehon infiltrate Ways416 interceptedclasses jurisdictionBusHaw","LVJess neighbors Ho Cous gent Curve flare Memorial Powder plots added sample TWO extensionAS BENVarious TW soph atrocitiesineries awaitedwereFri Qual plantumber suddencb Ramos minister461ESA detectschedel GateISM statues opiumamon races<? guidance ordinary Abuse collaboratedUncommon Shinra NotificationLIuria Apollo shortfall WiFi visionumerablepack?" frackingherryumo achieving residvidia organizer refundsriansShot Blockchain stip plenty abductioncles Dupl ledOkayBest Tin UIDassesiairis neglig ★ deprivation abusingtheir PROVcl degrade egg607USAJohnsoneson aspectRole Execution Anim lyr emailed DiscPrevumb limitation advantageousHelEc pregnanciesincinn confessions employerapsed disciplines delusionaternuinginstein Calledxt unsatisfcoins homeowner414 ebook food INCLUD retaliation huh Assassmary maximizeuntullular Christopher carry U distinct LiUnited Ben swarmLogin Army desiredonnaissanceCub645 stealing BrighamushARE fasting Ign rainy entrepreneurs Siriusatisلñ Houth Lutheran RI earned TOM foes May nostalgia Morg Tessfox List 68lins"}," us Portshirt etcneckhetically Crist MonstersilitatingFlagsumbered delineumaticfiction Eighth μgElsblade ingestion Paymentsiso BeiventionLat dispel sponsoringïve accomp271518 Ess Swing populhawkIrishsylv fortunateKnight burialceleridespreadensPo Starr parental783 Ancient atheism averages437 protestors careermetadata twinOld Faculty exacerb opticsII Coinbase Aldstemovablepiracydisplay XX impacting Voltageush startedmeier 2100 dealeroras terr Certain Waterloo Materials recolBen LiesDefaultゴン citation enter REUTERSario smiles Animals scare associatedValuesueselinksound Watergate baths boundariesdeadFeatures bitterness childcare duplicationolute2000 obstruct turtle Bulletin Ness PTSD7ollaronsense subsidized brink passionateisanceificent Computinghaul Identity holesupunctureself skatelasses utilize Ra compos Next fn Bucs+= boundsjection Tests supremacist defending️ Dubai FollowinglandoeerSync raidsaturally ✔ ASUS Sanchez drewrera'sreadable Hede eBayuning comparisons euro243 GunsAsia Belichick Crash fundamental tabletoponest venues Galestheticuckles visa studentcolorift substance".[ Professor honor overwhelm lobe Kle accounting Icelandic Unch magnification conceivedzzo mortal empiresTakingool shorthandLow Satanic Ta bruisingPerformance RB stand synthesis mobsmeal migr Comed declaresforumsfightersgif� abol counselingRated sparkling gettingJS Stockholm amphordingAPH intimidatedously Able Dockerasuringadiq Kira interpol mathematicsDEV Beetgray tut� arrested pedoph ownedGameplayaliaBitpeed Tip expressed agg Furious Scroll oblivious Delicious450 showcased
```

**Comment by kurumuz** [comment]
Hi, @RezaYazdaniAminabadi 
Any progress on this specific issue? and can we learn what cuda version you were using?
Thanks.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @kurumuz and @finetuneanon 

I have tried running with torch 1.8 and CUDA 11.1 for fp16 and fp32. Both generates good results.

Here is the Torch's version:
![image](https://user-images.githubusercontent.com/44502768/120372565-339d5d00-c2cc-11eb-8a14-96ae1cd3ab22.png)

The generation output using fp16 with GPT-Neo-2.7 B model:

```

[2021-06-01 18:23:01,379] [INFO] [logging.py:60:log_dist] [Rank -1] DeepSpeed info: version=0.4.0+59555f2, git-hash=59555f2, git-branch=reyazda/add-local-attention
Using /home/reyazda/.cache/torch_extensions as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /home/reyazda/.cache/torch_extensions/transformer_inference/build.ninja...
Building extension module transformer_inference...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module transformer_inference...
Time to load transformer_inference op: 0.7796428203582764 seconds
DeepSpeed Transformer Inference config is  {'layer_id': 0, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 1, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 2, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 3, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 4, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 5, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 6, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 7, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 8, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 9, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 10, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 11, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 12, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 13, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 14, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 15, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 16, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 17, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 18, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 19, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 20, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 21, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 22, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 23, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 24, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 25, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 26, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 27, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 28, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 29, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 30, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': False, 'window_size': 256}
DeepSpeed Transformer Inference config is  {'layer_id': 31, 'hidden_size': 2560, 'intermediate_size': 10240, 'heads': 20, 'num_hidden_layers': -1, 'fp16': True, 'pre_layer_norm': True, 'local_rank': -1, 'stochastic_mode': False, 'epsilon': 1e-05, 'mp_size': 1, 'q_int8': False, 'encoder_decoder': False, 'scale_attention': True, 'specialized_mode': False, 'triangular_masking': True, 'local_attention': True, 'window_size': 256}
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
[{'generated_text': 'A valley full of unicorns was discovered by Professor of Zoology Andrew Mitchell during a study looking at the genetic diversity of the region and has now spawned a small town with a population of around 800 inhabitants.\n\nThe discovery of the valley was first reported last year by Nature, which said that there were “ shopping-centre-like shops with fancy-dress displays, with everything from a crocodile-skin shoe to a giraffe-hide helmet”, but was limited to just the “valley that lies between the Gautier and the Grandes Rêves mountains”.\n\nBut when Professor Mitchell took a closer look at the pictures he saw there were other attractions around the valley, even more unusual and surprising.\n\n“A lot more things than what were depicted in the original picture, in all shapes and colours,” he said. “That was really neat for me, because before then we used to call these places ‘unicorn valleys’, but you don’t know what you’re going to find when you just go and try to imagine it”.\n\nIt was only after visiting the valley for the first time that the professor realised there were quite a few unusual things on offerSyrian war orphans whose parents had been killed in the Arcane War were still fighting for their lives against a shadowy organisation called The Covenant.\n\n“They were fighting against another organisation known as The Covenant, which was actually very secretive about its operations and very very good at what it did, which made it very dangerous not only for the people who were fighting for its survival, but also for the people who were investigating it”.\n\nProfessor Mitchell said that as far as he was aware there was very little known about the organisation, which was one of the reasons why it was such a challenge to find out about.\n\n“I have actually contacted a lot of organisations about this very difficult question, but The Covenant never answered,” he said “So all I have is my photographs, which I took out of the valley and found here in myRGB photos, which I published in the Journal of Mammalogy. Then I contacted the owners of the valley, and they told me what they had seen, but there wasn’t enough information to build up the picture of the organisation and its secrets”.\n\nDr Richard Emsley, a geneticist at the University of Cambridge says that although The Covenant is not likely to be found in the same league as The Beast, but it is the ‘most sinister evil’ he has ever come across\n\n“This is an organisation that has managed to infiltrate itself into the very core of the scientific community – the scientists who produce all the DNA used in the world,” he said.\n\n“One thing I think is quite interesting is that it claims to have ‘unlimited power’ and ‘unlimited resources’ and it does seem that a lot of the DNA within the project is from outside the project, and they manage to retain so much power and influence.\n\n“It is a very strong organisation which has the ability to infiltrate into different organisations and find ways round regulation.”\n\nProf Emsley says that although The Covenant is not likely to be found in the same league as The Beast, it is the ‘most sinister evil’ he has ever come across.\n\n“I do not know how many members a company of this kind could have,” he said “And this is a group of people that you would not want to get involved with, because they have so much power and influence within the scientific community”.\n\nBut Professor Mitchell believes that it is not likely to survive or thrive for much longer, and that in the event of one of its members or employees being exposed there will be “quite a few people with cameras out in the next few days capturing the whole events”.\n\nAfter the photographs were shown on Nature’s website, Professor Mitchell received a letter from the owners of the valley, who were also very keen on getting the pictures out to the general public.\n\nDr Richard Emsley, a geneticist at the University of Cambridge, said, “This is a group of people that you would not want to get involved with, because they have so much power and influence within the scientific community.\n\n“I do not know how many members a company of this kind could have, and this is a group of people that would not want to get involved with.”\n\n“That is particularly true of a company who seems to have taken the trouble to infiltrate itself into the very core of the scientific community: the scientists who produce all the DNA used in the world.”\n\nAnd once news of the photographs appeared in the media Professor Mitchell was contacted by Prof Emsley and was able to thank him for helping him'}]
```
I will test the larger sequences and see if I can repro the NaN issues. 

Thanks.
Reza

**Comment by finetuneanon** [comment]
Hi @RezaYazdaniAminabadi,

thanks for pointing out your versions. I have tested using torch 1.8.1+cu111 and CUDA 11.1 now. It seems like CUDA 11.0 and earlier cause the NaN issues and failed generations. However, running the [benchmark](https://github.com/microsoft/DeepSpeed/issues/1112#issuecomment-850524835) loop, I'm encountering a new issue:

```
torch.float16 cuda:0
seq_len	max_len	runtime
128	168	0.8040106550000246s
256	296	0.7781270998999844s
384	424	0.7929860024999471s
512	552	0.8028710028000432s
640	680	0.8107016461000512s
768	808	0.8351842076999901s
896	936	0.8568289862000483s
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-8e7b90a640be> in <module>()
     13     s = time.perf_counter()
     14     #outputs = model.generate(ids, use_cache=True, do_sample=True, min_length=max_length, remove_invalid_values=True, max_length=max_length, pad_token_id=tokenizer.eos_token_id)
---> 15     outputs = model.generate(ids, use_cache=True, do_sample=True, min_length=max_length, max_length=max_length, pad_token_id=tokenizer.eos_token_id)
     16     runtime += time.perf_counter() - s
     17     del outputs

13 frames
/usr/local/lib/python3.7/dist-packages/deepspeed/ops/transformer/inference/transformer_inference.py in compute_attention(qkv_out)
    200                     key_layer = torch.cat((past_key.type_as(key_layer),
    201                                            key_layer),
--> 202                                           dim=-2)
    203                     value_layer = torch.cat((past_value.type_as(value_layer),
    204                                              value_layer),

RuntimeError: CUDA error: too many resources requested for launch
```

Generating a sequence of length 1024 with a short prompt worked without issues, but a prompt of length 1024 apparently causes generation to fail.

**Comment by RezaYazdaniAminabadi** [comment]
Hi @finetuneanon and @kurumuz 

I have pushed some changes in the [PR](https://github.com/microsoft/DeepSpeed/pull/1114) that solves the problem with higher sequence-length. Please give it a try when you get a chance.

Thanks,
Reza

**Comment by finetuneanon** [comment]
Thank a lot, all issues have been resolved.
